src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * worker_shard_visibility.c
+ *
+ * Implements the functions for hiding shards on the Citus MX
+ * worker (data) nodes. This is mostly required for
+ *
+ *  * Copyright (c) 2018, Citus Data, Inc.
+ */
+
+#include "postgres.h"
+
+#include "catalog/index.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_class.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/worker_protocol.h"
+#include "distributed/worker_shard_visibility.h"
+#include "nodes/nodeFuncs.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+
+/* Config variable managed via guc.c */
+bool OverrideTableVisibility = true;
+
+static bool RelationIsAKnownShard(Oid shardRelationId);
+static Node * ReplaceTableVisibleFunctionWalker(Node *inputNode);
+
+PG_FUNCTION_INFO_V1(citus_table_is_visible);
+PG_FUNCTION_INFO_V1(relation_is_a_known_shard);
+
+
+/*
+ * relation_is_a_known_shard a wrapper around RelationIsAKnownShard(), so
+ * see the details there. The function also treats the indexes on shards
+ * as if they were shards.
+ */
+Datum
+relation_is_a_known_shard(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+
+	CheckCitusVersion(ERROR);
+
+	PG_RETURN_BOOL(RelationIsAKnownShard(relationId));
+}
+
+
+/*
+ * citus_table_is_visible aims to behave exactly the same with
+ * pg_table_is_visible with only one exception. The former one
+ * returns false for the relations that are known to be shards.
+ */
+Datum
+citus_table_is_visible(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	char relKind = '\0';
+
+	CheckCitusVersion(ERROR);
+
+	/*
+	 * We don't want to deal with not valid/existing relations
+	 * as pg_table_is_visible does.
+	 */
+	if (!SearchSysCacheExists1(RELOID, ObjectIdGetDatum(relationId)))
+	{
+		PG_RETURN_NULL();
+	}
+
+	if (RelationIsAKnownShard(relationId))
+	{
+		/*
+		 * If the input relation is an index we simply replace the
+		 * relationId with the corresponding relation to hide indexes
+		 * as well. See RelationIsAKnownShard() for the details and give
+		 * more meaningful debug message here.
+		 */
+		relKind = get_rel_relkind(relationId);
+		if (relKind == RELKIND_INDEX)
+		{
+			ereport(DEBUG1, (errmsg("skipping index \"%s\" since it belongs to a shard",
```

Comment:
Shall we make it `DEBUG2`, what do you think @mtuncer ?

Reply:
DEBUG2 might be better.